README
Beacony
=======
Based on AX25 Apps 0.0.8-rc4.
I removed a lot of stuff and what's behind is this program.
It listens to the AX25 interface and poops out beacons.
And that's all it does!

There's a script here that will `write` the output of the
beacon to your terminal which is the author's intended
usage case:
    sudo beacony/beacony | python beaconwriter.py

To build:
autoreconf -fi
./configure
make

-Jeff AG7EW, <jeff@baitis.net>; 3/5/2019


debian/changelog
ax25-apps (0.0.8-rc4-2build1) xenial; urgency=medium

  * No-change rebuild for ncurses6 transition.

 -- Matthias Klose <doko@ubuntu.com>  Sun, 07 Feb 2016 09:39:24 +0000

ax25-apps (0.0.8-rc4-2) unstable; urgency=medium

  * debian/control:
   - Change architecture to linux-any.
   - Removed inactive uploaders.

 -- Iain R. Learmonth <irl@debian.org>  Mon, 14 Sep 2015 00:29:00 +0100

ax25-apps (0.0.8-rc4-1) unstable; urgency=medium

  * New upstream version.
  * debian/control:
   - Switch to debhelper >= 9
   - Removed redundant Build-Depends:
    * dpkg-dev
    * autoconf
    * automake
   - Removed inactive uploaders:
    * Jaime Robles
    * Hamish Moffat (by request)
   - Added myself to uploaders
   - Added Homepage
   - Added Vcs-* for pkg-hamradio Git
   - Bump Standards-Version to 3.9.6
  * debian/watch:
   - Added from sepwatch
  * debian/compat:
   - Bump to 9
  * debian/{config,postinst}:
   - Added "set -e"
  * Source format switched to 3.0 (quilt)

 -- Iain R. Learmonth <irl@debian.org>  Wed, 09 Sep 2015 15:57:29 +0100

ax25-apps (0.0.8-rc2+cvs20130510-4) unstable; urgency=medium

  * Bump Standards Version.
  * Fix section for ax25ipd manpage.
  * Fix def of unix2yapp() (Closes: #749606)
  * Minor spelling fixes.

 -- Colin Tuckley <colint@debian.org>  Mon, 08 Sep 2014 21:08:51 +0100

ax25-apps (0.0.8-rc2+cvs20130510-3) unstable; urgency=low

  * Move to dh_autoreconf 
  * Add build depends for autotools

 -- Patrick Ouellette <pouelle@debian.org>  Sat, 11 May 2013 12:26:12 -0400

ax25-apps (0.0.8-rc2+cvs20130510-2) unstable; urgency=low

  * Fix build failure on buildd (autoreconf call)

 -- Patrick Ouellette <pouelle@debian.org>  Sat, 11 May 2013 11:53:32 -0400

ax25-apps (0.0.8-rc2+cvs20130510-1) unstable; urgency=low

  * License changed to BSD license for listen/ripdump.c (Closes: #630893)
  * Updated standards version to 3.9.4
  * Added use of hardening flags to debian/rules for compilation

 -- Patrick Ouellette <pouelle@debian.org>  FrI, 10 May 2013 11:13:29 -0400

ax25-apps (0.0.8-rc2+cvs20120204-2) unstable; urgency=low

  * Fix "ax25ipd: fails to transmit packets with "assemble_kiss: dumped
    - control byte non-zero"" fixed in new upstream
    (Closes: #606338)
  * Fix pending l10n issues. Debconf translations:
  * Italian (Vincenzo Campanella).  Closes: #593666
  * Danish (Joe Hansen).  Closes: #627843
  * Polish (Michał Kułach).  Closes: #659128
  * Fix "Uses absolute path to call dpkg-statoverride" change check for
    - dpkg-statoverride from test to hash (Closes: #510947)

 -- Patrick Ouellette <pouelle@debian.org>  Sun, 26 Feb 2012 20:46:29 -0500

ax25-apps (0.0.8-rc2+cvs20120204-1) unstable; urgency=low

  * new upstream

 -- Patrick Ouellette <pouelle@debian.org>  Sat, 04 Feb 2012 22:03:45 -0500

ax25-apps (0.0.6-16.2) unstable; urgency=low

  * Non-maintainer upload.
  * Bump debherlper compatibility level from deprecated 4 to 7
  * As a consequence, replace deprecated "dh_clean -k" by "dh_prep"
  * Add ${misc:Depends} to binary package dependencies to deal with
    dependencies triggerred by the use of debhelper
  * Explicitly point to the GPL-2 document in debian/copyright
  * Explicitly use 1.0 source format
  * No longer hardcode dpkg-statoverride path in postinst
  * Fix pending l10n issues. Debconf translations:
    - Russian (Yuri Kozlov).  Closes: #533876
    - Vietnamese (Clytie Siddall).  Closes: #579224

 -- Christian Perrier <bubulle@debian.org>  Sun, 25 Apr 2010 07:49:02 +0200

ax25-apps (0.0.6-16.1) unstable; urgency=low

  * Retiring - remove myself from the uploaders list.

 -- Joop Stakenborg <pa3aba@debian.org>  Tue, 10 Nov 2009 10:10:10 +0000

ax25-apps (0.0.6-16) unstable; urgency=low

  * Also rename the listen manual page to axlisten. Closes: #447315.
  * Adapt description in the control file.
  * Fix axlisten manual page, so we refer to axcall.
  * Do not call dh_installman.

 -- Joop Stakenborg <pa3aba@debian.org>  Sat, 01 Dec 2007 20:57:33 +0100

ax25-apps (0.0.6-15) unstable; urgency=low

  * Add missing 'missing' and 'depcomp'.
  * Run aclocal, automake autoconf and libtoolize again.
  * Listen is now called axlisten (closes: #447315), call is called
    axcall. Names were to generic. Adopt manpages.
  * ax25ipd/Makefile.am: use AM_CFLAGS instead of CFLAGS.
  * debian/menu: new menu section, 'Amateur Radio'.
  * Lintian clean.

 -- Joop Stakenborg <pa3aba@debian.org>  Sun, 11 Nov 2007 19:40:53 +0100

ax25-apps (0.0.6-14.1) unstable; urgency=low

  * Non-maintainer upload to fix pending l10n issues.
  * Debconf translations:
    - Convert files to UTF-8
    - Portuguese. Closes: #389649
    - Galician. Closes: #412342

 -- Christian Perrier <bubulle@debian.org>  Mon, 26 Feb 2007 07:31:59 +0100

ax25-apps (0.0.6-14) unstable; urgency=low

  * Remove build-dependency on autoconf, automake and libtool
    (closes: #376334)
  * Tidy up description.
  * Upgrade to standards-revision 3.7.2.
  * Upgrade to debhelper revision 4.

 -- Hamish Moffatt <hamish@debian.org>  Mon,  3 Jul 2006 08:42:54 +1000

ax25-apps (0.0.6-13) unstable; urgency=low

  * Close an already fixed bug. Closes: #321497.
  * Update config.guess, config.sub. Closes: #356363.

 -- Joop Stakenborg <pa3aba@debian.org>  Sun, 19 Mar 2006 22:00:24 +0100

ax25-apps (0.0.6-12) unstable; urgency=low

  * Changed debconf dependency to "debconf | debconf-2.0"
    to allow use with cdebconf
  * Updated standards-revision to 3.6.2

 -- Hamish Moffatt <hamish@debian.org>  Fri,  2 Sep 2005 09:14:18 +1000

ax25-apps (0.0.6-11) unstable; urgency=low

  * adds --update to postinst dpkg-statoverride call.  Closes #321497.

 -- Patrick A. Ouellette <pouelle@debian.org>  Mon,  8 Aug 2005 23:05:35 -0400

ax25-apps (0.0.6-10) unstable; urgency=low

  * Vietnamese translations for debconf. Closes: #313139.

 -- Joop Stakenborg <pa3aba@debian.org>  Mon, 11 Jul 2005 21:33:33 +0200

ax25-apps (0.0.6-9) unstable; urgency=low

  * Changed maintainer to debian-hams@lists.debian.org

beacony/beacony.c
#include <sys/types.h>
#include <netinet/in.h>
#include <sys/ioctl.h>
#include <netdb.h>

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <curses.h>
#include <signal.h>
#include <errno.h>

#include <sys/socket.h>
#include <net/if.h>
#ifdef __GLIBC__
#include <net/ethernet.h>
#else
#include <linux/if_ether.h>
#endif
#include <netax25/ax25.h>
#include <netax25/axconfig.h>

#include <config.h>
#include "beacony.h"

int timestamp;
static int sigint;
static int sock;

void display_timestamp(void)
{
	time_t timenowx;
	struct tm *timenow;

	time(&timenowx);
	timenow = localtime(&timenowx);

	lprintf(T_TIMESTAMP, "%02d:%02d:%02d", timenow->tm_hour,
		timenow->tm_min, timenow->tm_sec);
}

static void handle_sigint(int signal) 
{
	sigint++;
	close(sock);	/* disturb blocking recvfrom  */
	sock = -1;
}

#define ASCII		0
#define HEX 		1
#define READABLE	2

#define BUFSIZE		1500

int main(int argc, char **argv)
{
	unsigned char buffer[BUFSIZE];
	int dumpstyle = ASCII;
	int size;
	int s;
	char *port = NULL, *dev = NULL;
	struct sockaddr sa;
	socklen_t asize = sizeof(sa);
	struct ifreq ifr;
	int proto = ETH_P_AX25;
	int exit_code = EXIT_SUCCESS;

	timestamp = 0;

	while ((s = getopt(argc, argv, "8achip:rtv")) != -1) {
		switch (s) {
		case 'a':
			proto = ETH_P_ALL;
			break;
		case 'c':
			color = 1;
			break;
		case 'p':
			port = optarg;
			break;
		case 't':
			timestamp = 1;
			break;
		case 'v':
			printf("listen: %s\n", VERSION);
			return 0;
		case ':':
			fprintf(stderr,
				"listen: option -p needs a port name\n");
			return 1;
		case '?':
			fprintf(stderr,
				"Usage: listen [-a] [-c] [-p port] [-t] [-v]\n");
			return 1;
		}
	}

	if (ax25_config_load_ports() == 0)
		fprintf(stderr, "listen: no AX.25 port data configured\n");

	if (port != NULL) {
		if ((dev = ax25_config_get_dev(port)) == NULL) {
			fprintf(stderr, "listen: invalid port name - %s\n",
				port);
			return 1;
		}
	}

	if ((sock = socket(PF_PACKET, SOCK_PACKET, htons(proto))) == -1) {
		perror("socket");
		return 1;
	}

	if (color) {
		color = initcolor();	/* Initialize color support */
		if (!color)
			printf("Could not initialize color support.\n");
	}

	setservent(1);

	while (!sigint) {
		asize = sizeof(sa);

		signal(SIGINT, handle_sigint);
		signal(SIGTERM, handle_sigint);
		if ((size =
		     recvfrom(sock, buffer, sizeof(buffer), 0, &sa,
			      &asize)) == -1) {
			/*
			 * Signals are cared for by the handler, and we
			 * don't want to abort on SIGWINCH
			 */
			if (errno == EINTR) {
				refresh();
				continue;
			} else if (!(errno == EBADF && sigint)) {
				perror("recv");
				exit_code = errno;
			}
			break;
		}
		signal(SIGINT, SIG_DFL);
		signal(SIGTERM, SIG_DFL);
		if (sock == -1 || sigint)
			break;

		if (dev != NULL && strcmp(dev, sa.sa_data) != 0)
			continue;

		if (proto == ETH_P_ALL) {
			strcpy(ifr.ifr_name, sa.sa_data);
			signal(SIGINT, handle_sigint);
			signal(SIGTERM, handle_sigint);
			if (ioctl(sock, SIOCGIFHWADDR, &ifr) == -1) {
				if (!(errno == EBADF && sigint)) {
					perror("SIOCGIFHWADDR");
					exit_code = errno;
					break;
				}
			}
			signal(SIGINT, SIG_DFL);
			signal(SIGTERM, SIG_DFL);
			if (sock == -1 || sigint)
				break;
			if (ifr.ifr_hwaddr.sa_family == AF_AX25) {
				ki_dump(buffer, size, dumpstyle);
			}
		} else {
			ki_dump(buffer, size, dumpstyle);
		}
		if (color)
			refresh();
	}
	if (color)
		endwin();

	return exit_code;
}

static void baitisj_dump(unsigned char *data, int length) {
	unsigned char buf[1024];
	unsigned char* src=data;
	unsigned char* dst=buf;
        memset(buf, 0, sizeof(buf));
	while (*src != '\0' && src < (data + length) && dst+2 < buf+sizeof(buf)) {
		*dst = *src;
		if (*src == '\n') {
			*dst = '\\';
			++dst;
			*dst = 'n';
		}
		++src;
		++dst;
	}
	lprintf(T_DATA, "%s\n", buf);
}

void data_dump(unsigned char *data, int length, int dumpstyle)
{
	baitisj_dump(data, length);
}



beacony/ax25dump.c
/* AX25 header tracing
 * Copyright 1991 Phil Karn, KA9Q
 */
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include "beacony.h"

#define	LAPB_UNKNOWN	0
#define	LAPB_COMMAND	1
#define	LAPB_RESPONSE	2

#define	SEG_FIRST	0x80
#define	SEG_REM		0x7F

#define	PID_SEGMENT	0x08
#define	PID_ARP		0xCD
#define	PID_NETROM	0xCF
#define	PID_IP		0xCC
#define	PID_X25		0x01
#define	PID_TEXNET	0xC3
#define	PID_FLEXNET	0xCE
#define	PID_OPENTRAC	0x77
#define	PID_NO_L3	0xF0

#define	I		0x00
#define	S		0x01
#define	RR		0x01
#define	RNR		0x05
#define	REJ		0x09
#define	U		0x03
#define	SABM		0x2F
#define	SABME		0x6F
#define	DISC		0x43
#define	DM		0x0F
#define	UA		0x63
#define	FRMR		0x87
#define	UI		0x03
#define	PF		0x10
#define	EPF		0x01

#define	MMASK		7

#define	HDLCAEB		0x01
#define	SSID		0x1E
#define	REPEATED	0x80
#define	C		0x80
#define	SSSID_SPARE	0x40
#define	ESSID_SPARE	0x20

#define	ALEN		6
#define	AXLEN		7

#define	W		1
#define	X		2
#define	Y		4
#define	Z		8

static int ftype(unsigned char *, int *, int *, int *, int *, int);

#define NDAMA_STRING ""
#define DAMA_STRING " [DAMA]"

/*
 * 1: fm AG7EW to KG7OEM ctl UI^ pid=F0(Text) len 17
*/

/* FlexNet header compression display by Thomas Sailer t.sailer@alumni.ethz.ch */

/* Dump an AX.25 packet header */
void ax25_dump(unsigned char *data, int length, int hexdump)
{
	char heading[100];
	char tmp[15];
	char tmp2[15];
	int ctlen, nr, ns, pf, pid, seg, type, end, cmdrsp, extseq;
	char *dama;
	memset(heading, 0, sizeof(heading));

	/* check for FlexNet compressed header first; FlexNet header compressed packets are at least 8 bytes long */
	if (length < 8) {
		/* Something wrong with the header */
		lprintf(T_ERROR, "AX25: bad header!\n");
		return;
	}
	if (data[1] & HDLCAEB) {
		/* this is a FlexNet compressed header */
		tmp[6] = tmp[7] = extseq = 0;
		tmp[0] = ' ' + (data[2] >> 2);
		tmp[1] = ' ' + ((data[2] << 4) & 0x30) + (data[3] >> 4);
		tmp[2] = ' ' + ((data[3] << 2) & 0x3c) + (data[4] >> 6);
		tmp[3] = ' ' + (data[4] & 0x3f);
		tmp[4] = ' ' + (data[5] >> 2);
		tmp[5] = ' ' + ((data[5] << 4) & 0x30) + (data[6] >> 4);
		data += 7;
		length -= 7;
	} else {
		pax25(tmp, data + AXLEN);
                pax25(tmp2, data);
		snprintf(heading, sizeof(heading), "%s\t%s\t", tmp, tmp2);
		end = (data[AXLEN + ALEN] & HDLCAEB);
		data += (AXLEN + AXLEN);
		length -= (AXLEN + AXLEN);
		if (!end) {
			while (!end) {
				/* Print digi string */
				end = (data[ALEN] & HDLCAEB);
				data += AXLEN;
				length -= AXLEN;
			}
		}
	}

	if (length == 0)
		return;

	ctlen = ftype(data, &type, &ns, &nr, &pf, extseq);

	data += ctlen;
	length -= ctlen;

	if (type == I || type == UI) {
		/* Decode I field */
		if (length > 0) {	/* Get pid */
			pid = *data++;
			length--;

			if (timestamp)
				display_timestamp();

			if (pid == PID_SEGMENT) {
				length--;

				if (seg & SEG_FIRST) {
					pid = *data++;
					length--;
				}
			}

			switch (pid) {
				case PID_NO_L3:
					printf("%s", heading);
					data_dump(data, length, hexdump);
					break;
				default:
					break;
			}
		}
	} 
}

char *pax25(char *buf, unsigned char *data)
{
	int i, ssid;
	char *s;
	char c;

	s = buf;

	for (i = 0; i < ALEN; i++) {
		c = (data[i] >> 1) & 0x7F;

		if (!isalnum(c) && c != ' ') {
			strcpy(buf, "[invalid]");
			return buf;
		}

		if (c != ' ')
			*s++ = c;
	}

	if ((ssid = (data[ALEN] & SSID)) != 0)
		sprintf(s, "-%d", ssid >> 1);
	else
		*s = '\0';

	return (buf);
}

static int ftype(unsigned char *data, int *type, int *ns, int *nr, int *pf,
		 int extseq)
{
	*ns = *nr = 0;				/* To avoid warnings  */

	if (extseq) {
		if ((*data & 0x01) == 0) {	/* An I frame is an I-frame ... */
			*type = I;
			*ns = (*data >> 1) & 127;
			data++;
			*nr = (*data >> 1) & 127;
			*pf = *data & EPF;
			return 2;
		}
		if (*data & 0x02) {
			*type = *data & ~PF;
			*pf = *data & PF;
			return 1;
		} else {
			*type = *data;
			data++;
			*nr = (*data >> 1) & 127;
			*pf = *data & EPF;
			return 2;
		}
	} else {
		if ((*data & 0x01) == 0) {	/* An I frame is an I-frame ... */
			*type = I;
			*ns = (*data >> 1) & 7;
			*nr = (*data >> 5) & 7;
			*pf = *data & PF;
			return 1;
		}
		if (*data & 0x02) {	/* U-frames use all except P/F bit for type */
			*type = *data & ~PF;
			*pf = *data & PF;
			return 1;
		} else {	/* S-frames use low order 4 bits for type */
			*type = *data & 0x0F;
			*nr = (*data >> 5) & 7;
			*pf = *data & PF;
			return 1;
		}
	}
}


beacony/kissdump.c
/* Tracing routines for KISS TNC
 * Copyright 1991 Phil Karn, KA9Q
 */
#include <stdio.h>
#include "beacony.h"

#define	PARAM_DATA	0
#define	PARAM_TXDELAY	1
#define	PARAM_PERSIST	2
#define	PARAM_SLOTTIME	3
#define	PARAM_TXTAIL	4
#define	PARAM_FULLDUP	5
#define	PARAM_HW	6
#define	PARAM_RETURN	15	/* Should be 255, but is ANDed with 0x0F */

void ki_dump(unsigned char *data, int length, int hexdump)
{
	int type;
	int val;

	type = data[0] & 0xf;

	if (type == PARAM_DATA) {
		ax25_dump(data + 1, length - 1, hexdump);
		return;
	}

	val = data[1];

	switch (type) {
	case PARAM_TXDELAY:
		lprintf(T_KISS, "TX Delay: %lu ms\n", val * 10L);
		break;
	case PARAM_PERSIST:
		lprintf(T_KISS, "Persistence: %u/256\n", val + 1);
		break;
	case PARAM_SLOTTIME:
		lprintf(T_KISS, "Slot time: %lu ms\n", val * 10L);
		break;
	case PARAM_TXTAIL:
		lprintf(T_KISS, "TX Tail time: %lu ms\n", val * 10L);
		break;
	case PARAM_FULLDUP:
		lprintf(T_KISS, "Duplex: %s\n",
			val == 0 ? "Half" : "Full");
		break;
	case PARAM_HW:
		lprintf(T_KISS, "Hardware %u\n", val);
		break;
	case PARAM_RETURN:
		lprintf(T_KISS, "RETURN\n");
		break;
	default:
		lprintf(T_KISS, "code %u arg %u\n", type, val);
		break;
	}
}


beacony/utils.c
/*
 * Copyright 1996, 1997 Heikki Hannikainen OH7LZB <oh7lzb@sral.fi>
 *
 * Portions and ideas (like the ibm character mapping) from
 *	Tomi Manninen OH2BNS <oh2bns@sral.fi>
 */

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <curses.h>
#include <netdb.h>
#include <netinet/in.h>
#include <string.h>

#include "beacony.h"

int color = 0;			/* Colorized? */
int sevenbit = 1;		/* Are we on a 7-bit terminal? */
int ibmhack = 0;		/* IBM mapping? */

/* mapping of IBM codepage 437 chars 128-159 to ISO latin1 equivalents
 * (158 and 159 are mapped to space)
 */

unsigned char ibm_map[32] = {
	199, 252, 233, 226, 228, 224, 229, 231,
	234, 235, 232, 239, 238, 236, 196, 197,
	201, 230, 198, 244, 246, 242, 251, 249,
	255, 214, 220, 162, 163, 165, 32, 32
};

/*
 *	Printf in Technicolor (TM) (available in selected theatres only)
 */

void lprintf(int dtype, char *fmt, ...)
{
	va_list args;
	char str[1024];
	unsigned char *p;
	chtype ch;

	va_start(args, fmt);
	vsnprintf(str, 1024, fmt, args);
	va_end(args);

	if (color) {
		for (p = str; *p != '\0'; p++) {
			ch = *p;

			if (sevenbit && ch > 127)
				ch = '.';

			if ((ch > 127 && ch < 160) && ibmhack)
				ch = ibm_map[ch - 128] | A_BOLD;
			else if ((ch < 32) && (ch != '\n'))
				ch = (ch + 64) | A_REVERSE;

			if ((dtype == T_ADDR) || (dtype == T_PROTOCOL)
			    || (dtype == T_AXHDR) || (dtype == T_IPHDR)
			    || (dtype == T_ROSEHDR) || (dtype == T_PORT)
			    || (dtype == T_TIMESTAMP))
				ch |= A_BOLD;

			ch |= COLOR_PAIR(dtype);

			addch(ch);
		}
	} else {
		for (p = str; *p != '\0'; p++)
			if ((*p < 32 && *p != '\n')
			    || (*p > 126 && *p < 160 && sevenbit))
				*p = '.';
		if (fputs(str, stdout) == -1)
			exit(1);
		if (fflush(stdout) == -1)
			exit(1);
	}
}

int initcolor(void)
{
	if (!has_colors())
		return 0;
	initscr();		/* Start ncurses */
	start_color();		/* Initialize color support */
	refresh();		/* Clear screen */
	noecho();		/* Don't echo */
	wattrset(stdscr, 0);	/* Clear attributes */
	scrollok(stdscr, TRUE);	/* Like a scrolling Stone... */
	leaveok(stdscr, TRUE);	/* Cursor position doesn't really matter */
	idlok(stdscr, TRUE);	/* Use hardware ins/del of the terminal */
	nodelay(stdscr, TRUE);	/* Make getch() nonblocking */

	/* Pick colors for each type */
	init_pair(T_PORT, COLOR_GREEN, COLOR_BLACK);
	init_pair(T_DATA, COLOR_WHITE, COLOR_BLACK);
	init_pair(T_ERROR, COLOR_RED, COLOR_BLACK);
	init_pair(T_PROTOCOL, COLOR_CYAN, COLOR_BLACK);
	init_pair(T_AXHDR, COLOR_WHITE, COLOR_BLACK);
	init_pair(T_IPHDR, COLOR_WHITE, COLOR_BLACK);
	init_pair(T_ADDR, COLOR_GREEN, COLOR_BLACK);
	init_pair(T_ROSEHDR, COLOR_WHITE, COLOR_BLACK);
	init_pair(T_TIMESTAMP, COLOR_YELLOW, COLOR_BLACK);
	init_pair(T_KISS, COLOR_MAGENTA, COLOR_BLACK);
	init_pair(T_BPQ, COLOR_MAGENTA, COLOR_BLACK);
	init_pair(T_TCPHDR, COLOR_BLUE, COLOR_BLACK);
	init_pair(T_FLEXNET, COLOR_BLUE, COLOR_BLACK);
	init_pair(T_OPENTRAC, COLOR_YELLOW, COLOR_BLACK);


	return 1;
}

char *servname(int port, char *proto)
{
	struct servent *serv;
	static char str[16];

	if ((serv = getservbyport(htons(port), proto)))
		strncpy(str, serv->s_name, 16);
	else
		snprintf(str, 16, "%i", port);

	return str;
}
